appc/appconnector.go

```diff
@@ -222,83 +232,126 @@ nextAnswer:
 			}
 			continue
 		}
-		if h.Type != dnsmessage.TypeA && h.Type != dnsmessage.TypeAAAA {
+
+		switch h.Type {
+		case dnsmessage.TypeCNAME, dnsmessage.TypeA, dnsmessage.TypeAAAA:
+		default:
 			if err := p.SkipAnswer(); err != nil {
 				return
 			}
 			continue
+
 		}
 
-		domain := h.Name.String()
+		domain := strings.TrimSuffix(strings.ToLower(h.Name.String()), ".")
 		if len(domain) == 0 {
-			return
-		}
-		domain = strings.TrimSuffix(domain, ".")
-		domain = strings.ToLower(domain)
-		e.logf("[v2] observed DNS response for %s", domain)
-
-		e.mu.Lock()
-		addrs, ok := e.domains[domain]
-		// match wildcard domains
-		if !ok {
-			for _, wc := range e.wildcards {
-				if dnsname.HasSuffix(domain, wc) {
-					e.domains[domain] = nil
-					ok = true
-					break
-				}
-			}
+			continue
 		}
-		e.mu.Unlock()
 
-		if !ok {
-			if err := p.SkipAnswer(); err != nil {
+		if h.Type == dnsmessage.TypeCNAME {
+			res, err := p.CNAMEResource()
+			if err != nil {
 				return
 			}
+			cname := strings.TrimSuffix(strings.ToLower(res.CNAME.String()), ".")
+			if len(cname) == 0 {
+				continue
+			}
+			mak.Set(&cnameChain, cname, domain)
 			continue
 		}
 
-		var addr netip.Addr
 		switch h.Type {
 		case dnsmessage.TypeA:
 			r, err := p.AResource()
 			if err != nil {
 				return
 			}
-			addr = netip.AddrFrom4(r.A)
+			addr := netip.AddrFrom4(r.A)
+			mak.Set(&addressRecords, domain, append(addressRecords[domain], addr))
 		case dnsmessage.TypeAAAA:
 			r, err := p.AAAAResource()
 			if err != nil {
 				return
 			}
-			addr = netip.AddrFrom16(r.AAAA)
+			addr := netip.AddrFrom16(r.AAAA)
+			mak.Set(&addressRecords, domain, append(addressRecords[domain], addr))
 		default:
 			if err := p.SkipAnswer(); err != nil {
 				return
 			}
 			continue
 		}
-		if slices.Contains(addrs, addr) {
-			continue
-		}
-		for _, route := range e.controlRoutes {
-			if route.Contains(addr) {
-				// record the new address associated with the domain for faster matching in subsequent
-				// requests and for diagnostic records.
-				e.mu.Lock()
-				e.domains[domain] = append(addrs, addr)
-				e.mu.Unlock()
-				continue nextAnswer
+	}
+
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+nextRecord:
+	for domain, addrs := range addressRecords {
+		var isRouted bool
+
+		// starting from the domain that resolved to an address, find it, or any
+		// of the domains in the CNAME chain toward resolving it, that are
+		// routed domains, replacing domain with the routed domain name.
+	findRoutedDomain:
+		for {
+			_, isRouted = e.domains[domain]
+			if isRouted {
+				break findRoutedDomain
+			}
+
+			// match wildcard domains
+			for _, wc := range e.wildcards {
+				if dnsname.HasSuffix(domain, wc) {
+					e.domains[domain] = nil
+					isRouted = true
+					break findRoutedDomain
+				}
+			}
+
+			next, ok := cnameChain[domain]
+			if !ok {
+				break findRoutedDomain
 			}
+			domain = next
 		}
```

Comment:
Could we pull this into a function, and return a domain and a boolean to indicate a success? e.g. 

```go 
foundDomain, ok := e.findRoutedDomain(domain, cnameChain)
if !ok {
continue 
}
```
Then we could use the foundDomain where we call use 'domain' below currently, and avoid overriding the loop variable.

Reply:
done!